src/StreamSelectLoop.php

```diff
@@ -220,27 +213,117 @@ private function waitForStreamActivity($timeout)
+    {
+        if ($time === null) {
+            return null;
+        }
+        $fractional = fmod($time, 1);
+        $microseconds = round($fractional * self::MICROSECONDS_PER_SECOND);
+
+        return intval($microseconds);
+    }
+
+    /**
+     * Returns integer amount of nanoseconds in $time or null.
+     * The precision is 1 microsecond.
+     *
+     * @param float|null $time – time in seconds
+     *
+     * @return int|null
+     */
+    private static function getNanoseconds($time)
+    {
+        if ($time === null) {
+            return null;
+        }
+
+        return intval(self::getMicroseconds($time) * self::NANOSECONDS_PER_MICROSECOND);
+    }
+
+    /**
      * Emulate a stream_select() implementation that does not break when passed
      * empty stream arrays.
      *
      * @param array        &$read   An array of read streams to select upon.
      * @param array        &$write  An array of write streams to select upon.
-     * @param integer|null $timeout Activity timeout in microseconds, or null to wait forever.
+     * @param float|null   $timeout Activity timeout in seconds, or null to wait forever.
      *
      * @return integer|false The total number of streams that are ready for read/write.
      * Can return false if stream_select() is interrupted by a signal.
      */
     protected function streamSelect(array &$read, array &$write, $timeout)
     {
+        $seconds = self::getSeconds($timeout);
+        $microseconds = self::getMicroseconds($timeout);
+        $nanoseconds = self::getNanoseconds($timeout);
+
         if ($read || $write) {
-            $except = null;
+            $except = [];
 
-            // suppress warnings that occur, when stream_select is interrupted by a signal
-            return @stream_select($read, $write, $except, $timeout === null ? null : 0, $timeout);
+            return $this->doSelectStream($read, $write, $except, $seconds, $microseconds);
         }
 
-        $timeout && usleep($timeout);
+        $this->sleep($seconds, $nanoseconds);
 
         return 0;
     }
+
+    /**
+     * Proxy for built-in stream_select method.
+     *
+     * @param array $read
+     * @param array $write
+     * @param array $except
+     * @param int|null $seconds
+     * @param int|null $microseconds
+     *
+     * @return int
+     */
+    protected function doSelectStream(array &$read, array &$write, array &$except, $seconds, $microseconds = null)
+    {
+        // suppress warnings that occur, when stream_select is interrupted by a signal
+        return @stream_select($read, $write, $except, $seconds, $microseconds);
+    }
+
+    /**
+     * Sleeps for $seconds and $nanoseconds.
+     *
+     * @param int|null $seconds
+     * @param int|null $nanoseconds
+     */
+    protected function sleep($seconds, $nanoseconds = 0)
+    {
+        if ($seconds > 0 || $nanoseconds > 0) {
+            time_nanosleep($seconds, $nanoseconds);
+        }
```

Comment:
I'd like to see explicit handling of `$seconds`or `$nanoseconds`being `null` here.